File: C++/13. Roman to Integer.cpp
class Solution {
public:

    int romanToInt(string s) {
        //define roman as integer
        int table[128] = {0};
        table['I'] = 1;
        table['V'] = 5;
        table['X'] = 10;
        table['L'] = 50;
        table['C'] = 100;
        table['D'] = 500;
        table['M'] = 1000;

        // if s.size()<=1
        if (s.size() == 0)
            return 0;
        if (s.size() == 1)
            return table[s[0]];

        int temp = 0;
        int res = 0;
        for (int i = 0; i < s.size();) {
            // the roman_before < roman-after like IV=4
            if (i + 1 < s.size() && table[s[i]] < table[s[i + 1]]) {
                temp = table[s[i + 1]] - table[s[i]];
                i += 2;
            } else { //normal
                temp = table[s[i]];
                i++;
            }
            res += temp;
        }
        return res;
    }
};

File: 算法/Python/13. Roman to Integer.py
class Solution:
    def romanToInt(self, s):
        """
        :type s: str
        :rtype: int
        """
        # define a roman dict
        dictionary = {'I': 1, 'V': 5, 'X': 10,
                      'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        if len(s) == 0:
            return 0
        if len(s) == 1:
            return dictionary[s]

        res = 0
        i = 0
        while i < len(s):
            # if s[i]<s[i+1] like IV=4
            if i+1 < len(s) and dictionary[s[i]] < dictionary[s[i+1]]:
                res += dictionary[s[i+1]] - dictionary[s[i]]
                i += 2
            else:
                res += dictionary[s[i]]
                i += 1
        return res
